machine/utils/aio.py

```diff
@@ -0,0 +1,103 @@
+# -*- coding: utf-8 -*-
+
+import asyncio
+import concurrent
+from functools import partial
+from typing import Any, Callable, Coroutine, List, Optional, Sequence, Tuple, Type
+
+
+async def join(tasks: Sequence[Coroutine]) -> List[Any]:
+    """ Execute all of the coroutines, returning a list of responses
+        or exceptions.
+    """
+
+    async def wrapper(future, idx):
+        try:
+            output = await future
+            return idx, output
+        except Exception as err:
+            return idx, err
+
+    output = [None] * len(tasks)
+    tasks = [wrapper(future, idx) for idx, future in enumerate(tasks, start=0)]
+
+    for future in asyncio.as_completed(tasks):
+        idx, result = await future
+        output[idx] = result
+
+    return output
+
+
+async def split(tasks: Sequence[Coroutine]) -> Tuple[List[Any], List[Type[Exception]]]:
+    """ Execute all of the coroutines, returning a list of responses
+        and a list of exceptions.
+    """
+
+    results = await join(tasks)
+
+    return_values = []
+    exceptions = []
+
+    for res in results:
+        if isinstance(res, Exception):
+            exceptions.append(res)
+        else:
+            return_values.append(res)
+
+    return return_values, exceptions
+
+
+def run_coro(coro: Coroutine, loop: Optional[asyncio.AbstractEventLoop] = None) -> Any:
```

Comment:
Part of me wishes this were renamed to indicate that it blocks until the coro finishes. I'm torn because I suppose it should be "obvious" wdyt?